modules/module-mongodb-storage/src/storage/MongoBucketStorage.ts

```diff
@@ -289,19 +268,22 @@ export class MongoBucketStorage
     });
   }
 
-  async getActiveCheckpoint(): Promise<storage.ActiveCheckpoint> {
-    const doc = await this.db.sync_rules.findOne(
-      {
-        state: storage.SyncRuleState.ACTIVE
-      },
-      {
-        sort: { _id: -1 },
-        limit: 1,
-        projection: { _id: 1, last_checkpoint: 1, last_checkpoint_lsn: 1 }
-      }
-    );
+  async getActiveStorage(): Promise<MongoSyncBucketStorage | null> {
+    const content = await this.getActiveSyncRulesContent();
+    if (content == null) {
+      return null;
+    }
 
-    return this.makeActiveCheckpoint(doc);
+    // It is important that this instance is cached.
+    // Not for the instance construction itself, but to ensure that internal caches on the instance
+    // are re-used properly.
+    if (this.activeStorageCache?.group_id == content.id) {
+      return this.activeStorageCache;
+    } else {
+      const instance = this.getInstance(content);
+      this.activeStorageCache = instance;
```

Comment:
Should we `dispose` the previous `activeStorageCache` before assigning a new reference?

Reply:
I'm still trying to wrap my head around disposing of SyncRulesBucketStorage. Especially when used in the sync API, the lifetime of the instance is not well-defined. Even when the instance is not in the `activeStorageCache` anymore, there could still be references to it being used.

In the case of the sync APIs, it is fine if references are still being used after disposing, since disposing only affect listeners, and listeners are not relevant for the sync API.

@stevensJourney What is the practical effect of disposing when used for replication? What will happen if we never dispose the instance?

Reply:
Discussed offline: My primary concern was with the custom write checkpoint module, which registered listeners on the sync bucket storage. I initially thought the callbacks for those listeners might hold references to the sync bucket storage, potentially preventing garbage collection. However, after further review, it appears that we don't track any references in those listeners.